product-scenarios/code-coverage/code-coverage.sh

```diff
@@ -0,0 +1,152 @@
+
+            echo $(tinkerer_curl_commands ${agent_name[i]} "cd\t$VAR_REM_DIR/repository/logs/jacoco;\tzip\t-r\tjacoco.zip\t.")
+
+            #Generate key
+            echo "Generate a key in node ${agent_name[i]}"
+
+            echo $(tinkerer_curl_commands ${agent_name[i]} "mkdir\t-p\t$VAR_REM_DIR/../keys")
+            echo $(tinkerer_curl_commands ${agent_name[i]} "ssh-keygen\t-b\t2048\t-t\trsa\t-f\t$VAR_REM_DIR/../keys/deploy.key\t-q\t-N\t''")
+
+            #Add key to authorized_keys
+            echo "Add key of node ${agent_name[i]} to the authorized_keys in client side"
+
+            echo `tinkerer_curl_commands ${agent_name[i]} "cat\t$VAR_REM_DIR/../keys/deploy.key.pub"` | \
+            sed -nE 's/.*"response":"(.*)","completed.*/\1/p' >> /home/ubuntu/.ssh/authorized_keys
+
+            #Copy file to client side
+            echo "Copy file to the client side from ${agent_name[i]}"
+            mkdir -p ${HOME}/code-coverage/resources/instance$((i+1))
+
+            #Give permission to .key file
+            tinkerer_curl_commands ${agent_name[i]} "sudo\tchmod\t400\t$VAR_REM_DIR/../keys/deploy.key"
+
+            echo $(tinkerer_curl_commands ${agent_name[i]} \
+            "scp\t-o\tStrictHostKeyChecking=no\t-i\t$VAR_REM_DIR/../keys/deploy.key\t$VAR_REM_DIR/repository/logs/jacoco/jacoco.zip\t$user@$ipaddress:$HOME/code-coverage/resources/instance$((i+1))")
+            sleep 30
+
+            echo "Extract the copied coverage artifact file of ${agent_name[i]}"
+            unzip ${HOME}/code-coverage/resources/instance$((i+1))/jacoco.zip \
+            -d ${HOME}/code-coverage/resources/instance$((i+1))/jacoco
+
+            sleep 40
+
+            i=$((i+1))
+
+        done < "${HOME}"/code-coverage/resources/agent_name_list.txt
+
+        #Execute code-coverage POM and generate coverage reports
+        echo "Generate coverage reports from coverage artifacts"
+        mvn clean install -f ${HOME}/code-coverage/pom.xml
+
+        #Copy Code Coverage Reports
+        echo "Copy code coverage reports to the output directory"
+        cp -r ${HOME}/code-coverage/target/scenario-code-coverage ${OUTPUT_DIR}
+
+    fi
+}
+
+#=== FUNCTION ==================================================================
+# NAME: testgrid tinkerer curl commands
+# DESCRIPTION: Curl command to perform shell commands on server side through testgrid tinkerer
+# PARAMETER 1: agent_name
+# PARAMETER 2: shell_command which need to execute on server sides
+#===============================================================================
+function tinkerer_curl_commands(){
+
+    agent_name=$1
+    shell_command=$2
+
+    curl -X POST "${VAR_TINKERER_ENDPOINT}"/api/test-plan/"${VAR_TEST_PLAN_ID}"/agent/"${agent_name}"/operation \
+           -H 'content-type: application/json' \
+           -d '{"code":"SHELL", "request":"'$shell_command'"}' \
+           -u "${VAR_TINKERER_USERNAME}":"${VAR_TINKERER_PASSWORD}"
+
+     sleep 30
+}
```

Comment:
Shall we add an empty line in the end. Some editors don't like it when this is not available.
```suggestion
}

```